### ISSUE TITLE
storage: occasional indefinite storage.Writer hangs

### ISSUE THREAD

**Issue by joshuaseaton** [googleapis/google-cloud-go#1624]
I have storage write logic that effectively looks like
```
w := bucket.Object(name).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
w.ChunkSize = chunkSize
w.MD5 = expectedChecksum
var err error
for {
  if _, err = io.CopyN(w, r, chunkSize); err != nil {
    break
  }
}
// EOFs and 412s are okay. 412s are okay as this routine might have
// lost a race with another routine that uploaded the associated
// object first, thus failing the above precondition.
if err != nil && err != io.EOF && !string.Contains(err.Error(), "Error 412") {  {
  w.Close()
  return err
}
return w.Close()
```
Occasionally I witness indefinite hangs. What is the issue?

## Client

Storage

## Describe Your Environment
GCE instance
*cores*: 16
*cpu*: x64 | x86 | x86-64-Haswell_GCE | x86-64-avx2
*gpu*: none
*kvm*: true
*machine_type*: n1-standard-16
*os*: Debian | Debian-9.5 | Linux
*ssd*: true

**Comment by joshuaseaton** [comment]
CC: frankyn@

**Comment by eliben** [comment]
@joshuaseaton can you reliably reproduce this issue with the latest release? If yes, could you do a goroutine dump (`Ctrl+\`) while it hangs?

**Comment by tbpg** [comment]
Talked to @tritone offline. We haven't been able to reproduce the issue. Closing this as it's been a week, but let us know if it's still happening and we can reopen.

**Comment by mfederw** [comment]
I am seeing the writer.Close() hang on the second close.  I put a first object into the bucket do a writer.Close().  It shows up in the bucket.  The second object write attempts the write and then w.Close() hangs.  I have the goroutine dump here:

[godump.txt](https://github.com/googleapis/google-cloud-go/files/5626493/godump.txt)

The hung Close is at:
goroutine 82 [chan receive]:
cloud.google.com/go/storage.(*Writer).Close(0xc0000f6900, 0x0, 0x0)
        /dwshome/mike/go/pkg/mod/cloud.google.com/go/storage@v1.12.0/writer.go:230 +0x92

I guess the usefull other routine is:
goroutine 147 [select]:
cloud.google.com/go/storage.(*Writer).monitorCancel(0xc0000f6900)
        /dwshome/mike/go/pkg/mod/cloud.google.com/go/storage@v1.12.0/writer.go:240 +0xc7
created by cloud.google.com/go/storage.(*Writer).open
        /dwshome/mike/go/pkg/mod/cloud.google.com/go/storage@v1.12.0/writer.go:106 +0x334

Any help would be appreciated.

**Comment by tymeshifter** [comment]
I've stumbled upon this problem too so +1 from me for reopening it. I've only reproduced it in unit tests where I am using https://github.com/fsouza/fake-gcs-server

What I've localized it to, is the interaction between preconditions and http response status.

In more detail for this specific case `fake-gcs-server` doesn't support `ifGenerationMatch=X` where X>0  condition, so it returns `http.StatusNotImplemented`.  This makes `Writer.Close` to hang indefinitely. I've confirmed that returning `http.StatusPreconditionFailed` resolves it.

So my assumption is that there is a bug in what client expect to see in response status and this makes it never complete a request.
